### PR TITLE
Adds ability to set required and optional checks

### PR DESCRIPTION
**Pull request by sebastienvas**
Redoing #7457 as it takes forever to load.

Ref: https://goo.gl/cd9MEu which was discussed with @fejta @cjwagner @BenTheElder 

Instead of getting required checks from the github api, it will use the branch protection configuration and infere the required context from it. We are also adding a optional options to Prow Presubmits.

Config would look like this:
```yaml
...
tide:
  ...
  context_options:
    policy:
      # Use branch protection options to define required and optional contexts
      from-branch-protection: true
      # Treat unknown contexts as optional
      skip-unknown-contexts: true
    orgs:
      org:
        repo:
          branch:
            policy:
              from-branch-protection: false
              required-contexts:
              - "required_test"
              optional-contexts:
              - "optional_test"
  
```
#7140

### CODE REVIEW

File: prow/config/branch_protection.go [hunk 1]
```diff
@@ -142,19 +146,24 @@ func jobRequirements(jobs []Presubmit, branch string, after bool) []string {
...
+			optional = append(optional, j.Context)
 		}
 		// Check which children require contexts
-		required = append(required, jobRequirements(j.RunAfterSuccess, branch, true)...)
+		r, o := jobRequirements(j.RunAfterSuccess, branch, true)
+		required = append(required, r...)
+		optional = append(optional, o...)
 	}
-	return required
+	return
```
Comment (stevekuznetsov) [inline prow/config/branch_protection.go]: -1 to named returns

File: prow/config/branch_protection.go [hunk 2]
```diff
@@ -142,19 +146,24 @@ func jobRequirements(jobs []Presubmit, branch string, after bool) []string {
...
 }
 
-func branchRequirements(org, repo, branch string, presubmits map[string][]Presubmit) []string {
+// BranchRequirements returns required and optional presubmits prow jobs for a given org, repo branch.
+func BranchRequirements(org, repo, branch string, presubmits map[string][]Presubmit) (required, optional []string) {
 	p, ok := presubmits[org+"/"+repo]
 	if !ok {
-		return nil
+		return
```
Comment (stevekuznetsov) [inline prow/config/branch_protection.go]: -1 to named returns

File: prow/config/branch_protection_test.go [hunk 3]
```diff
@@ -298,11 +298,11 @@ func TestJobRequirements(t *testing.T) {
 	}
 
 	for _, tc := range cases {
-		masterActual := jobRequirements(tc.config, "master", false)
+		masterActual, _ := jobRequirements(tc.config, "master", false)
```
Comment (stevekuznetsov) [inline prow/config/branch_protection_test.go]: extend the test to test both?

File: prow/cmd/branchprotector/README.md [hunk 4]
```diff
@@ -22,6 +22,8 @@ branch-protection:
           # Ensure that the extra-process-followed github status context passes.
           # In addition, adds any required prow jobs (aka always_run: true)
           require-contexts: ["extra-process-followed"]
+          # Tell Tide to merge when unspecified contexts are pending or failing
+          skip-unknown-contexts: true
```
Comment (stevekuznetsov) [inline prow/cmd/branchprotector/README.md]: Configuration for `tide` actions should not live in the branch protector's config.

File: prow/tide/contextregister.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+*/
+
+package tide
+
+import (
+	"sync"
+
+	"github.com/shurcooL/githubql"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/prow/config"
```
Comment (stevekuznetsov) [inline prow/tide/contextregister.go]: nit: indent of this repo should be in separate block at bottom

File: prow/tide/contextregister.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+)
+
+type contextChecker interface {
+	// ignoreContext tells whether a context is optional.
+	ignoreContext(context Context) bool
```
Comment (stevekuznetsov) [inline prow/tide/contextregister.go]: how about `IsOptional(context Context) bool` ?

File: prow/tide/contextregister.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+	// missingRequiredContexts tells if required contexts are missing from the list of contexts provided.
+	missingRequiredContexts([]Context) []Context
```
Comment (stevekuznetsov) [inline prow/tide/contextregister.go]: nit: name the param? or don't name the one above

File: prow/tide/contextregister.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+// newExpectedContext creates a Context with Expected state.
+// This should not only be used when contexts are missing.
```
Comment (stevekuznetsov) [inline prow/tide/contextregister.go]: >  // This should not only be used when contexts are missing. 

What does this mean?

File: prow/tide/contextregister.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+// newContextRegister instantiates a new contextRegister and register tide as optional by default
+// and uses Prow Config to find optional and required tests, as well as skipUnknownContexts.
+func newContextRegister(skipUnknownContexts bool) *contextRegister {
+	r := &contextRegister{
+		required:            sets.NewString(),
+		optional:            sets.NewString(),
+		skipUnknownContexts: skipUnknownContexts,
+	}
+	r.registerOptionalContexts(statusContext)
```
Comment (stevekuznetsov) [inline prow/tide/contextregister.go]: where is the `statusContext` identifier coming from here?

File: prow/tide/contextregister.go [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+// and uses Prow Config to find optional and required tests, as well as skipUnknownContexts.
+func newContextRegisterFromPolicy(policy *config.Policy) *contextRegister {
+	r := newContextRegister(false)
+	if policy != nil {
+		if policy.SkipUnknownContexts != nil {
+			r.skipUnknownContexts = *policy.SkipUnknownContexts
+		}
+		if policy.Protect != nil && *policy.Protect {
+			r.registerRequiredContexts(policy.Contexts...)
+		} else if r.skipUnknownContexts {
```
Comment (stevekuznetsov) [inline prow/tide/contextregister.go]: If the policy is not to protect but we are skipping unknown contexts we want to require the contexts the policy has listed to not protect on? What?

File: prow/tide/contextregister.go [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+	return Context{
+		Context:     githubql.String(c),
+		State:       githubql.StatusStateExpected,
+		Description: githubql.String(""),
+	}
+}
+
+// contextRegister implements contextChecker and allow registering of required and optional contexts.
+type contextRegister struct {
+	lock                sync.RWMutex
```
Comment (stevekuznetsov) [inline prow/tide/contextregister.go]: Why do we need this?

File: prow/tide/contextregister.go [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+	r.registerOptionalContexts(optional...)
+	return r, nil
+}
+
+// ignoreContext checks whether a context can be ignored.
+// Will return true if
+// - context is registered as optional
+// - required contexts are registered and the context provided is not required
+// Will return false otherwise. Every context is required.
+func (r *contextRegister) ignoreContext(c Context) bool {
```
Comment (stevekuznetsov) [inline prow/tide/contextregister.go]: Shouldn't we also ignore the `tide` context regardless of whether or not it's required?

File: prow/tide/contextregister.go [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+		return true
+	}
+	return false
+}
+
+// missingRequiredContexts discard the optional contexts and only look of extra required contexts that are not provided.
+func (r *contextRegister) missingRequiredContexts(contexts []Context) []Context {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	if r.required.Len() == 0 {
```
Comment (stevekuznetsov) [inline prow/tide/contextregister.go]: Won't the `sets.Difference()` make this happen automatically? No need for this check

File: prow/tide/tide.go [hunk 14]
```diff
@@ -357,15 +357,22 @@ func (sc *statusController) setStatuses(all, pool []PullRequest) {
...
 		}
-
-		wantState, wantDesc := expectedStatus(queriesByRepo, pr, poolM)
+		cr, err := newContextRegisterFromConfig(
+			string(pr.Repository.Owner.Login),
+			string(pr.Repository.Name),
+			string(pr.BaseRef.Name),
+			sc.ca.Config())
+		if err != nil {
+			log.WithError(err).Error("Getting branch protection")
```
Comment (stevekuznetsov) [inline prow/tide/tide.go]: nit: bad error

File: prow/tide/tide.go [hunk 15]
```diff
@@ -569,34 +576,37 @@ func toSimpleState(s kube.ProwJobState) simpleState {
...
+// unsuccessfulContexts determines which contexts from the list that we care about are
+// failed. For instance, we do not care about our own context.
+// If the branchProtection is set to only check for required checks, we will skip
+// all non-required tests. If required tests are missing from the list, they will be
+// added to the list of failed contexts.
+func unsuccessfulContexts(contexts []Context, cc contextChecker) []Context {
 	var failed []Context
 	for _, ctx := range contexts {
-		if string(ctx.Context) == statusContext {
+		if cc.ignoreContext(ctx) {
```
Comment (stevekuznetsov) [inline prow/tide/tide.go]: this seems like a regression if we mark the `tide` status required

File: prow/tide/tide.go [hunk 16]
```diff
@@ -944,7 +954,7 @@ func (c *Controller) presubmitsByPull(sp subpool) (map[int]sets.String, error) {
 	}()
 
 	for _, ps := range c.ca.Config().Presubmits[sp.org+"/"+sp.repo] {
-		if ps.SkipReport || !ps.RunsAgainstBranch(sp.branch) {
+		if !ps.ContextRequired() || !ps.RunsAgainstBranch(sp.branch) {
```
Comment (stevekuznetsov) [inline prow/tide/tide.go]: Why are we changing this functionality? If we want to do this, let's do it in a separate PR?

File: prow/tide/tide_test.go [hunk 17]
```diff
@@ -1582,6 +1589,123 @@ func TestSync(t *testing.T) {
...
+	failure := string(githubql.StatusStateFailure)
+	testCases := []struct {
+		name             string
+		passing          bool
+		optional         []string
+		config           *config.Policy
+		combinedContexts map[string]string
+	}{
+		{
+			name:    "required checks disabled should not impact - passing",
```
Comment (stevekuznetsov) [inline prow/tide/tide_test.go]: these names are a bit confusing

File: prow/tide/tide_test.go [hunk 18]
```diff
@@ -1582,6 +1589,123 @@ func TestSync(t *testing.T) {
...
+			passing: true,
+			config: &config.Policy{
```
Comment (stevekuznetsov) [inline prow/tide/tide_test.go]: @fejta is it even valid config to have a policy with explicitly listed required checks but have the policy be set to not enforce? why?

File: prow/config/branch_protection.go (hunk 1 above)
Comment (kargakis) [reply prow/config/branch_protection.go]: Agreed on not using named returns

File: prow/cmd/branchprotector/README.md (hunk 4 above)
Comment (kargakis) [reply prow/cmd/branchprotector/README.md]: +1, this should be part of the job spec

File: prow/config/branch_protection.go (hunk 1 above)
Comment (sebastienvas) [reply prow/config/branch_protection.go]: May I ask why ?

File: prow/cmd/branchprotector/README.md (hunk 4 above)
Comment (sebastienvas) [reply prow/cmd/branchprotector/README.md]: For prow jobs we already have the optional setting. For optional jobs that are not prowjob we need to define it somewhere. I had a global setting in tide configuration, but that was not practical. It makes sense for me to define it in branch protection since this option only works when branch protection is set and you might want to apply this globally, on a repo or on a branch which branch protection already does. I had discussed this with @cjwagner

File: prow/tide/tide.go (hunk 15 above)
Comment (sebastienvas) [reply prow/tide/tide.go]: cc already ignores tide

File: prow/tide/tide.go (hunk 16 above)
Comment (sebastienvas) [reply prow/tide/tide.go]: This is the goal of this PR

File: prow/cmd/branchprotector/README.md (hunk 4 above)
Comment (sebastienvas) [reply prow/cmd/branchprotector/README.md]: also skipunknown context does not have any meaning if we don't have a branch protection, meaning no jobs defined at all in Prow config since all are unknown.

File: prow/config/branch_protection.go (hunk 1 above)
Comment (stevekuznetsov) [reply prow/config/branch_protection.go]: I find almost universally they are harder to reason about and read through. They are almost nonexistant in the codebase as it is. Stats:

```
$ grep -rh func prow/ | grep -Po "(?<=\) \()[^\(\)]+(?=\) \{)" | tr ',' '\n' | awk '{print $2 }' | grep -v "^$" | wc -l 
19
$ grep -rh func prow/ | wc -l 
2084
```

File: prow/cmd/branchprotector/README.md (hunk 4 above)
Comment (sebastienvas) [reply prow/cmd/branchprotector/README.md]: So the original idea was to use the github required check to figure that out, but that was doing too many request to the github api. After a meeting with the fejta@, cjwagner@ and @BenTheElder it was decided to use branch protection instead. You don't have to actually enforce branch protection, you can just use the config and set protect to false, and then use that to define the required checks that you required for a PR to be merged. I don't see the point of duplicating those configs since they map back to protecting a branch. I think the Prow config is big and complicated enough.

File: prow/tide/tide.go (hunk 16 above)
Comment (sebastienvas) [reply prow/tide/tide.go]: The reason is that SkipReport = false does not mean the check is required. In our use case you might want to push a status on a PR for a test that is not required, this is the reason I added Optional a while back. Optional is false by default so it will not impact anything, but then if you define a prowjob with Optional set to true, you defintely want to skip it to merge a PR. So this is the reason I added this.

File: prow/tide/tide_test.go (hunk 17 above)
Comment (sebastienvas) [reply prow/tide/tide_test.go]: thanks for pointing that out. After reviewing some of the test I discovered some issues.

File: prow/config/branch_protection_test.go (hunk 3 above)
Comment (sebastienvas) [reply prow/config/branch_protection_test.go]: done

File: prow/config/branch_protection.go (hunk 2 above)
Comment (sebastienvas) [reply prow/config/branch_protection.go]: fixed

File: prow/cmd/branchprotector/README.md (hunk 4 above)
Comment (stevekuznetsov) [reply prow/cmd/branchprotector/README.md]: > You don't have to actually enforce branch protection, you can just use the config and set protect to false, and then use that to define the required checks that you required for a PR to be merged. 

Err, what? If you want to configure branch protection for statuses, do that. Then let `tide` honor required statuses. If you want to configure statuses for `tide`, do that, and don't touch the branch protection config (which you might have never even written since you don't use the branch protector). IDK. These seem like two entirely separate use cases that are being put in one place.

File: prow/config/branch_protection.go (hunk 1 above)
Comment (sebastienvas) [reply prow/config/branch_protection.go]: I actually discovered them in k8s codebase :) but right it may be confusing ! fixed.

File: prow/tide/contextregister.go (hunk 5 above)
Comment (sebastienvas) [reply prow/tide/contextregister.go]: done

File: prow/tide/tide.go (hunk 14 above)
Comment (sebastienvas) [reply prow/tide/tide.go]: fixed

File: prow/tide/tide_test.go (hunk 18 above)
Comment (sebastienvas) [reply prow/tide/tide_test.go]: I think there could multiple use cases. When you define policy at the repo or org level, and then you create a branch for testing and you don't want to protect it or if you just want to disable temporarily.

**Comment by sebastienvas** [comment]
PTAL

File: prow/tide/tide.go (hunk 16 above)
Comment (stevekuznetsov) [reply prow/tide/tide.go]: I understand they are not identical features, what I am saying is today we use `SkipReport` to mean "tide ignore" because whether or not it's required, if it's not reported, `tide` ignores it. As long as `SkipReport=true` continues to mean that `tide` ignores the job, but now in addition, `SkipReport=false Optional=true` works, that is OK.

**Comment by stevekuznetsov** [comment]
Please no merge commits. Rebase on `master`

File: prow/cmd/branchprotector/README.md (hunk 4 above)
Comment (sebastienvas) [reply prow/cmd/branchprotector/README.md]: I agree I am defining a tide options which will basically allow you to use branch protection or otherwise allow you to define your own if you don't want to use branch protection. If that option is not set, it will use combined status from github as it is today.

File: prow/tide/tide.go (hunk 16 above)
Comment (sebastienvas) [reply prow/tide/tide.go]: ok.

**Comment by sebastienvas** [comment]
@stevekuznetsov I think we got it. I think you are going to like this.

File: prow/tide/contextregister.go (hunk 6 above)
Comment (sebastienvas) [reply prow/tide/contextregister.go]: done

File: prow/tide/contextregister.go (hunk 7 above)
Comment (sebastienvas) [reply prow/tide/contextregister.go]: done

File: prow/tide/contextregister.go (hunk 8 above)
Comment (sebastienvas) [reply prow/tide/contextregister.go]: I guess It was a reminder for me to use it only a context was missing from a combined status. I removed the comment.

File: prow/tide/contextregister.go (hunk 11 above)
Comment (sebastienvas) [reply prow/tide/contextregister.go]: I had it since tide code is using multiple go routine, but since the object is created inside a funciton, the lock is overkill.

File: prow/tide/contextregister.go (hunk 9 above)
Comment (sebastienvas) [reply prow/tide/contextregister.go]: it is coming from tide.go (its the tide status). I know the name is not very well chosen, but I am trying to do the least amount of changes.

File: prow/tide/contextregister.go (hunk 12 above)
Comment (sebastienvas) [reply prow/tide/contextregister.go]: we do, that s tide status. its registered by default. This implementation could be use in another context, so I did not want to hard code in there.

File: prow/tide/contextregister.go (hunk 13 above)
Comment (sebastienvas) [reply prow/tide/contextregister.go]: it will but why bother doing all that when you already know the answer.

File: prow/tide/contextregister.go [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+func newExpectedContext(c string) Context {
+	return Context{
+		Context:     githubql.String(c),
+		State:       githubql.StatusStateExpected,
+		Description: githubql.String(""),
+	}
+}
+
+// contextRegister implements contextChecker and allow registering of required and optional contexts.
+type contextRegister struct {
```
[the remark this answers is not in the document]
Comment (sebastienvas) [reply prow/tide/contextregister.go]: I tend to agree with you, but that would create circular dependency and I would have to move the Context struct out as an example. Plus I would have to hard code the tide status in there. I am trying to keep this as the minimum amount of change. However, I don't think that the context checker interface is something that we need. The implementation is very general and straightforward. What do you think ?

[the remark this answers is not in the document]
Comment (cjwagner) [reply prow/tide/contextregister.go]: What is the circular dependency? The `contextRegister` is basically copying the fields from `TideContextPolicy` and defining new methods so I'd really prefer to condense these into one struct and just configure the packages and imports properly to avoid a circular dependency.

The tide status shouldn't need to be hard coded? It can be passed down from the `tide` package.

[the remark this answers is not in the document]
Comment (sebastienvas) [reply prow/tide/contextregister.go]: The circular deps is on the Context object defined in prow/tide/tide.go

File: prow/config/tide.go [hunk 20]
```diff
@@ -216,3 +248,72 @@ func (tq *TideQuery) Validate() error {
...
+	// Using Branch protection configuration
+	if options.FromBranchProtection != nil && *options.FromBranchProtection {
+		bp, err := c.GetBranchProtection(org, repo, branch)
+		if err != nil {
+			return TideContextPolicy{}, err
+		}
+		if bp == nil {
+			return TideContextPolicy{}, errors.New("branch protection is not set")
+		}
+		if bp.Protect == nil || !*bp.Protect || bp.RequiredStatusChecks == nil {
```
[the remark this answers is not in the document]
Comment (sebastienvas) [reply prow/config/tide.go]: done

File: prow/config/tide.go [hunk 21]
```diff
@@ -216,3 +248,72 @@ func (tq *TideQuery) Validate() error {
...
+
+func mergeTideContextPolicyConfig(a, b TideContextPolicy) TideContextPolicy {
+	mergeBool := func(a, b *bool) *bool {
+		if b == nil {
+			return a
+		}
+		return b
+	}
+	c := TideContextPolicy{}
+	c.FromBranchProtection = mergeBool(c.FromBranchProtection, b.FromBranchProtection)
```
[the remark this answers is not in the document]
Comment (sebastienvas) [reply prow/config/tide.go]: done

File: prow/config/tide.go [hunk 22]
```diff
@@ -216,3 +248,72 @@ func (tq *TideQuery) Validate() error {
...
+		}
+		if bp == nil {
+			return TideContextPolicy{}, errors.New("branch protection is not set")
+		}
+		if bp.Protect == nil || !*bp.Protect || bp.RequiredStatusChecks == nil {
+			return TideContextPolicy{}, errors.New("branch protection is invalid")
+		}
+		required.Insert(bp.RequiredStatusChecks.Contexts...)
+	}
+
```
[the remark this answers is not in the document]
Comment (sebastienvas) [reply prow/config/tide.go]: done and added tc

File: prow/tide/tide.go [hunk 23]
```diff
@@ -991,6 +1002,10 @@ func (c *Controller) syncSubpool(sp subpool) (Pool, error) {
 	if err != nil {
 		return Pool{}, fmt.Errorf("error determining required presubmits: %v", err)
 	}
+	cr, err := newContextRegisterFromConfig(sp.org, sp.repo, sp.branch, c.ca.Config())
+	if err != nil {
+		return Pool{}, fmt.Errorf("error parsing branch protection: %v", err)
```
[the remark this answers is not in the document]
Comment (sebastienvas) [reply prow/tide/tide.go]: done

File: prow/tide/contextregister.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (sebastienvas) [reply prow/tide/contextregister.go]: Done

**Comment by sebastienvas** [comment]
@cjwagner Please take a look

File: prow/config/tide.go [hunk 24]
```diff
@@ -216,3 +248,136 @@ func (tq *TideQuery) Validate() error {
...
+// RegisterOptionalContexts registers optional contexts
+func (cp *TideContextPolicy) RegisterOptionalContexts(c ...string) {
+	o := sets.NewString(cp.OptionalContexts...)
+	o.Insert(c...)
+	cp.OptionalContexts = o.List()
+}
+
+// RegisterRequiredContexts register required contexts.
+// Once required contexts are registered other contexts will be considered optional.
+func (cp *TideContextPolicy) RegisterRequiredContexts(c ...string) {
```
Comment (cjwagner) [inline prow/config/tide.go]: This func is no longer needed.

File: prow/config/tide.go [hunk 25]
```diff
@@ -216,3 +248,136 @@ func (tq *TideQuery) Validate() error {
...
+	}
+	var missingContexts []string
+	for c := range sets.NewString(cp.RequiredContexts...).Difference(existingContexts) {
+		missingContexts = append(missingContexts, c)
+	}
+	return missingContexts
+}
+
+// RegisterOptionalContexts registers optional contexts
+func (cp *TideContextPolicy) RegisterOptionalContexts(c ...string) {
```
Comment (cjwagner) [inline prow/config/tide.go]: It might make more sense to just pass the tide context name to GetTideContextPolicy and have that func make the tide context required. That is all that RegisterOptionalContexts is used for now.

Not a big deal either way.

Comment (sebastienvas) [reply prow/config/tide.go]: We can also just revert part of my change such that tide skip its own status. What do you think ?

**Comment by sebastienvas** [comment]
@cjwagner Please take a look. I ll be deploying this today for istio. Will keep you up to date.
